iree-prof-tools/iree-vis.cc

```diff
@@ -4,35 +4,153 @@
+#include "third_party/llvm-project/mlir/include/mlir/IR/BuiltinOps.h"
+#include "third_party/llvm-project/mlir/include/mlir/IR/Dialect.h"
 #include "third_party/llvm-project/mlir/include/mlir/IR/Operation.h"
 
 ABSL_FLAG(std::string, input_iree_file, "",
           "IREE MLIR text assembly file to read.");
-ABSL_FLAG(std::string, iree_phase, "executable-sources",
-          "IREE compilation phase of input IREE MLIR file.");
-
-void Walk(mlir::Operation* op, absl::string_view indent) {
-  LOG(INFO) << indent << op->getName().getStringRef().str();
-  std::string child_indent = absl::StrCat(indent, "  ");
-  op->walk([op, indent = absl::string_view(child_indent)](
-      mlir::Operation* nested_op) {
-    if (nested_op != op) {
-      Walk(nested_op, indent);
+ABSL_FLAG(std::string, output_json_file, "", "Graph JSON file to write.");
+ABSL_FLAG(std::string, function, "",
+          "Entrypoint function. If empty, pick one with most stream ops.");
+
+namespace iree_prof {
+namespace {
+
+absl::StatusOr<mlir::func::FuncOp> GetFuncWithMostStreams(
+    mlir::ModuleOp module) {
+  int max_num_streams = 0;
+  absl::StatusOr<mlir::func::FuncOp> result =
+      absl::NotFoundError("Can't find a func with streams");
+
+  module->walk([&](mlir::func::FuncOp func) {
+    int num_streams = 0;
+    func.walk([&num_streams](
+        mlir::iree_compiler::IREE::Stream::CmdExecuteOp exec) {
+      exec.walk([&num_streams](mlir::Operation* op) {
+        if (llvm::isa<mlir::iree_compiler::IREE::Stream::StreamDialect>(
+                op->getDialect())) {
+          ++num_streams;
+        }
+      });
+    });
+
+    if (num_streams > max_num_streams) {
+      max_num_streams = num_streams;
+      result = func;
+    }
+  });
+
+  if (result.ok()) {
+    LOG(INFO) << "Func with max streams = "  << result->getSymName().str()
+              << ", # of streams = " << max_num_streams;
+  }
+  return result;
+}
+
+absl::StatusOr<mlir::func::FuncOp> GetFuncWithName(mlir::ModuleOp module,
+                                                   absl::string_view name) {
+  absl::StatusOr<mlir::func::FuncOp> result = absl::NotFoundError(
+      absl::StrCat("Can't find a func of name \"", name, "\""));
+
+  module->walk([&](mlir::func::FuncOp func) -> mlir::WalkResult {
+    if (name == func.getSymName().str()) {
+      result = func;
+      return mlir::WalkResult::skip();
     }
+    return mlir::WalkResult::advance();
   });
+
+  if (result.ok()) {
+    LOG(INFO) << "Found a func with name = " << result->getSymName().str();
+  }
+  return result;
+}
+
+absl::Status AddNode(mlir::func::FuncOp func, graph::Graph& graph) {
+  graph.nodes.emplace_back();
+  graph::GraphNode& node = graph.nodes.back();
+  node.node_id = func.getSymName().str();
+  node.node_label = func.getSymName().str();
+  node.node_name = func.getSymName().str();
+  return absl::OkStatus();
+}
+
+absl::StatusOr<graph::GraphCollection> GetGraphCollection(
+    mlir::Operation* op,
```

Comment:
nit: we can directly use ModuleOp here and do the reinterpret cast in main().

Reply:
Done